[all] use =deleted for deprecation

damienmarchal opened the pull request:
Use ```= delete``` to make the deprecated function to fails at compilation


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
[all] Remove the macro SOFA_BEGIN_DEPRECATION_AS_ERROR

The use of = delete () is more readable and works on any system.
So I think replacing the macro with this is ok.

damienmarchal committed:
[all] remove all deprecated function.

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

fredroy commented on SofaKernel/modules/SofaCore/src/sofa/core/MultiMapping.h:
```diff
@@ -213,7 +213,7 @@ class MultiMapping : public BaseMapping
                           "to one or more mechanical states of type '" + std::string(TIn::Name()) + "'.");
             return false;
-        } else if (!LinkFromModels::CheckPaths( input, context )) {
+        } else if (!PathResolver::CheckPaths(context, LinkFromModels::DestType::GetClass(), input)) {
```
Look fine to me, except those lines which seems to not follow the topic of this PR

damienmarchal replied:
You are right, I restored it. The PathResolver will come next in its own PR.

damienmarchal committed:
Fred

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into xp-clean

fredroy commented on modules/SofaValidation/src/SofaValidation/EvalPointsDistance.h:
```diff
@@ -120,12 +120,12 @@ class EvalPointsDistance: public virtual sofa::core::objectmodel::BaseObject
         std::string object1 = arg->getAttribute("object1","@./");
         std::string object2 = arg->getAttribute("object2","@./");
-        if (!LinkMState::CheckPath(object1, context)) {
+        if (!sofa::core::PathResolver::CheckPath(context, LinkMState::DestType::GetClass(), object1)) {
```
Still issues with PathResolver here and there

damienmarchal replied:
Actually... they are deleted !!! So it was perfectly in line with the topic :)

But I restore the CheckPath because we depreciate them 1 week ago so ...waiting a bit more may be kind.

fredroy commented:
And you are sure to remove the Macro SOFA_BEGIN_DEPRECATION_AS_ERROR  ? 
It could be useful for delete/deprecate things other than methods (where we can use =delete)

damienmarchal commented:
@fredroy 
I don't know, I think that =delete is far cleaner for most of the cases and I think we have already so much work to do with the =delete that we should start with it ;) 
Nevertheless, as the one responsible for the macro I would prefer all developper to use it a lot but I doubt it will happen. 
I restored the macro so we can move forward and remove it if we see in few month that no one is using it.

damienmarchal committed:
WIP

damienmarchal committed:
WIP

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into xp-clean

fredroy commented:
[ci-build][with-all-tests][force-full-build]

guparan reviewed (changes requested):
TODO: update deprecation messages to clarify that those functions are now REMOVED.

damienmarchal committed:
Add macros in config.h.in to handle the different "deprecation" state.

The idea is that each deprecation should have its own macro in which we encode
the current status of the deprecation.

damienmarchal committed:
Use the introduced macro in our codebase

damienmarchal committed:
Add deprecation macro for "old" deprecation process.

damienmarchal committed:
Add missing copyValue in BaseData

guparan committed:
[SofaCore] CLEAN deprecated/delete mechanism

guparan commented:
I fixed all the wrong dates and cleaned the overall deprecation mechanism in https://github.com/CRIStAL-PADR/sofa/pull/5
Waiting for review @damienmarchal

damienmarchal commented on SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseLink.h:
```diff
@@ -108,8 +106,8 @@ class SOFA_CORE_API BaseLink
 
-    [[deprecated("2020-10-03: Deprecated since PR #1503. BaseLink cannot hold Data anymore. Use DataLink instead. Please update your code. ")]]
-    bool isDataLink() const { return false; }
+    SOFA_DEPRECATE_BASEDATA_API("Link cannot hold Data anymore. To update your code your code you need to use a DataLink.")
```
Should be SOFA_DEPRECATE_DATALINK.

damienmarchal commented on SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseLink.h:
```diff
@@ -131,16 +129,17 @@ class SOFA_CORE_API BaseLink
-    [[deprecated("2020-10-03: Deprecated since PR #1503. BaseLink cannot hold Data anymore. Use DataLink instead. Please update your code. ")]]
-    virtual BaseData* getLinkedData(std::size_t index=0) const = 0;
+
+    SOFA_DEPRECATE_BASEDATA_API("Link cannot hold Data anymore. To update your code your code you need to use a DataLink.")
```
Should be SOFA_DEPRECATE_DATALINK.

guparan committed:
[SofaCore] Define one macro per deprecation group

guparan committed:
[SofaCore] Better separation between name and description in deprecation macros

guparan committed:
[SofaCore] FIX and CLEAN deprecation macros

damienmarchal committed:
Merge pull request #5 from guparan/xp-clean

[SofaCore] CLEAN deprecated/delete mechanism

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

guparan committed:
[SofaCore] Use versions instead of dates for deprecation macros

guparan committed:
[SofaCore] Use versions instead of dates for deprecation macros (2)

damienmarchal commented:
@guparan  what is the requested change on this PR ?

guparan commented:
@damienmarchal https://github.com/CRIStAL-PADR/sofa/pull/6 :+1:

guparan committed:
Update config.h.in

guparan committed:
Update config.h.in

guparan commented:
# Deprecation macros

## The 2 base macros

**Do not use them directly in your code**, see "how to use them" below.

`SOFA_ATTRIBUTE_DEPRECATED(deprecateDate, removeDate, toFixMsg)`
To be used to trigger a deprecation warning. It is a simple `[[deprecated]]` interface.
Warns that something
&nbsp;&nbsp;&nbsp; - is DEPRECATED (so still usable) since `deprecateDate`
&nbsp;&nbsp;&nbsp; - will be removed on `removeDate`
and gives the toFixMsg instructions to fix the deprecation warning.
    
`SOFA_ATTRIBUTE_DISABLED(deprecateDate, disableDate, toFixMsg)`
To be used jointly with `= delete` to trigger an error. It is a flavored `[[deprecated]]` interface.
Warns that something  
&nbsp;&nbsp;&nbsp; - is DISABLED (so not usable anymore) since `disabledDate`
&nbsp;&nbsp;&nbsp; - was firstly deprecated on `deprecateDate`
and gives the toFixMsg instructions to fix the compilation error.


## How to use them

Create a new deprecation macro specifying the deprecation topic you are tackling.
This way, you won't have dates everywhere in your code.

## Example

### Deprecate

I want to rename something that will break my API. 
To ease the transition for my users, I will propose a deprecation period with a fine message by using a deprecation macro.
Latest release is v20.06, next release (currently under development) is v20.12, after that will come the v21.06.

In my config.h.in
```cpp
#define SOFA_ATTRIBUTE_DEPRECATED__MYDEPRECATIONTOPIC() \
    SOFA_ATTRIBUTE_DEPRECATED( \
        "v20.12 (PR#12345)", "v21.06", \
        "XXX must be renamed into YYY.")
```

In my code
```cpp
SOFA_ATTRIBUTE_DEPRECATED__MYDEPRECATIONTOPIC()
void myDeprecatedMethod();
```

Anyone using myDeprecatedMethod will get this **warning**: 
```text
warning: 'myDeprecatedMethod' is deprecated: It is still usable but has been DEPRECATED since v20.12 (PR#12345). You have until v21.06 to fix your code. XXX must be renamed into YYY.
```

### Disable

Now forward in time ...
Latest release is v20.12, next release (currently under development) is v21.06, after that will come the v21.12.
It is time to stop the deprecation period. To provide a clear message to my users, I will not just remove my deprecated method but I will disable it with `= delete` and use a deprecation macro.

In my config.h.in
```cpp
#define SOFA_ATTRIBUTE_DISABLED__MYDEPRECATIONTOPIC() \
    SOFA_ATTRIBUTE_DISABLED( \
        "v20.12 (PR#12345)", "v21.06 (PR#45678)", \
        "XXX must be renamed into YYY.")
```

In my code
```cpp
SOFA_ATTRIBUTE_DISABLED__MYDEPRECATIONTOPIC()
void myDeprecatedMethod() = delete;
```

Anyone using myDeprecatedMethod will get an **error** right after this warning: 
```text
warning: 'myDeprecatedMethod' is deprecated: It is not usable anymore because it has been DISABLED since v21.06 (PR#45678) after being deprecated on v20.12 (PR#12345). XXX must be renamed into YYY.
```

damienmarchal committed:
Merge pull request #6 from guparan/xp-clean

[SofaCore] Use versions instead of dates for deprecation macros

damienmarchal commented:
Ready, then go.
